File: packages/mirascript/src/vm/operations/convert.ts
import { toNumber, toString, toBoolean, toFormat } from '../../helpers/convert/index.js';
import type { VmAny } from '../types/index.js';
import { $AssertInit } from './common.js';

/** 转换为布尔值 */
export const $ToBoolean = (value: VmAny): boolean => {
    if (typeof value == 'boolean') return value;
    $AssertInit(value);
    return toBoolean(value, undefined);
};
/** 转换为字符串 */
export const $ToString = (value: VmAny): string => {
    if (typeof value == 'string') return value;
    $AssertInit(value);
    return toString(value, undefined);
};
/** 转换为数字 */
export const $ToNumber = (value: VmAny): number => {
    if (typeof value == 'number') return value;
    $AssertInit(value);
    return toNumber(value, undefined);
};
/** 格式化值 */
export const $Format = (value: VmAny, format: string | null): string => {
    $AssertInit(value);
    return toFormat(value, format);
};

File: crates/python/mirascript/_vm/operations/convert.py
from __future__ import annotations

from ..._helpers.convert import to_boolean, to_format, to_number, to_string
from ..types import VmAny
from .common import AssertInit


def ToBoolean(value: VmAny) -> bool:
    if isinstance(value, bool):
        return value
    AssertInit(value)
    return to_boolean(value)


def ToNumber(value: VmAny) -> float:
    if isinstance(value, float):
        return value
    AssertInit(value)
    return to_number(value)


def ToString(val: VmAny) -> str:
    if isinstance(val, str):
        return val
    AssertInit(val)
    if val is None:
        return ""
    return to_string(val)


def Format(val: VmAny, fmt: str | None = None) -> str:
    if isinstance(val, str):
        return val
    AssertInit(val)

    return to_format(val, fmt)
